streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1323,15 +1327,18 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
             ZIO.succeed(Right(exec.getEmit))
           case ChannelState.Effect(zio) =>
             zio *> interpret(exec.run().asInstanceOf[ChannelState[Env, OutErr]])
-          case r @ ChannelState.Read(upstream, onEffect, onEmit, onDone) =>
+          case r: ChannelState.Read[Env, OutErr] =>
             ChannelExecutor.readUpstream[Env, OutErr, OutErr, Either[OutDone, OutElem]](
-              r.asInstanceOf[ChannelState.Read[Env, OutErr]],
-              () => interpret(exec.run().asInstanceOf[ChannelState[Env, OutErr]]),
-              Exit.failCause
+              r = r,
+              onSuccess = () => interpret(exec.run().asInstanceOf[ChannelState[Env, OutErr]]),
```

Comment:
Can this function be moved into a `val`?
```suggestion
val onSuccess = () => interpret(exec.run().asInstanceOf[ChannelState[Env, OutErr]])
...
              onSuccess = onSuccess,
```

Reply:
May I ask why do you prefer this? 🤔

Reply:
To avoid the allocation for the function on every `Read` if the `val` can be defined next to the `setFinalizer`

Reply:
Thanks. Done :)